tests: cache managed Python installs

nooscraft opened the pull request:
Cache managed Python downloads (#16721)

konstin commented on crates/uv/tests/it/python_install.rs:
```diff
@@ -2718,18 +2679,19 @@ fn python_install_no_cache() {
         .arg("3.12")
-        .arg("--offline"), @r"
+        .arg("--offline")
+        .env(EnvVars::UV_PYTHON_CACHE_DIR, offline_cache.as_os_str()), @r"
```
Why are we using the cache dir in the offline test?

nooscraft commented:
> How fast are the tests before and after, does the cache work?

I ran `cargo test --package uv --test it python_install -- --nocapture` before and after. The very first one still takes about 42 s because we start with an empty cache, but every run after that drops into the 38–39 s range, managed downloads get reused and we stop pulling over the network each time.

nooscraft commented:
Thanks for the feedback, @zanieb! I want to make sure I understand correctly before pushing the changes

I've made with_python_download_cache() non-public and removed all explicit calls from test files (python_install.rs, python_find.rs, python_upgrade.rs, and pip_install.rs).

Since with_managed_python_dirs() now calls it internally, tests using .with_managed_python_dirs() will automatically get the cache enabled. Is this the intended behavior you were looking for?

You mentioned we need to exercise the code path where the variable is unset. I've changed it to:
```
pub fn without_python_download_cache(mut self) -> Self {
    let key: OsString = EnvVars::UV_PYTHON_CACHE_DIR.into();
    self.extra_env.retain(|(existing, _)| existing != &key);
    self
}
```

This removes the variable from extra_env entirely, rather than setting it to an empty directory. Is this what you meant by unsetting the variable, or did you mean something else?

(I want to confirm this correctly tests the "cache disabled" scenario vs just testing "empty cache directory".)

nooscraft commented:
@konstin @zanieb I’ve pushed the changes so it’s easier for you to review what has actually changed, instead of going in circles. If this isn’t the expected outcome, it’s possible I misunderstood the intent behind the original change request. If that’s the case, I’m happy to hand this over to someone with a better understanding of the test and cache logic.
